### ISSUE TITLE
HTML not converted 

### ISSUE THREAD

**Issue by HeenaAttar** [jstedfast/MailKit#640]
While replying to mail using mailkit the mail body is not send properly, HTML content is not forwarded as original, but for GMAIL it works perfectly, whereas for other servers like Hotmail or custom mails the mail body is not forwarded as orginal HTML body. Please consider here in this secnario if the receiving server is GMAIL then mail body is displayed properly, but if receiver server is different from GMAIL then contents are disturbed, also if sender is GMAIL and receiving is other server then contents are send as expected. Only happens if receiving and sender server is not GMAIL. Also note that this happens only when mail body contains HTML, images and CSS contents

**Comment by jstedfast** [comment]
You'll have to show me your code. It's really hard to follow what you are saying.

Also, I can 100% guarantee it's not a MimeKit or MailKit problem.

**Comment by HeenaAttar** [comment]
Hello Sir,

I am sharing my Reply Code using mailkit please have a look and let me know
where i have mistaken.

*CONTROLLER CODE:*

 *//Reply Mail*
        public ActionResult Reply(int Index, bool ReplyAllFlag,string
Foldername)
        {
            EmailDetails model = new EmailDetails();
            try
            {
                using (var client = new ImapClient())
                {
                    ServicePointManager.ServerCertificateValidationCallback
+= (o, c, ch, er) => true;
                    client.Connect(objSettings.IMAPHost,
objSettings.IMAPPort, true);
                    client.AuthenticationMechanisms.Remove("XOAUTH");
                    client.Authenticate(objSettings.UserName,
objSettings.Password);
                    model.FolderName = Foldername;
                    var mail = client.Inbox;
                    mail.Open(FolderAccess.ReadWrite);
                    model.InboxCount = client.Inbox.Count;
                    if ((client.Capabilities & (ImapCapabilities.SpecialUse
| ImapCapabilities.XList)) != 0)
                    {

                        if (model.FolderName == "Drafts")
                        {

                            mail = client.GetFolder(SpecialFolder.Drafts);
                        }
                        else if (model.FolderName == "Trash")
                        {
                            mail = client.GetFolder(SpecialFolder.Trash);
                        }
                        else if (model.FolderName == "Sent")
                        {
                            mail = client.GetFolder(SpecialFolder.Sent);
                        }
                        else if (model.FolderName == "Junk")
                        {
                            mail = client.GetFolder(SpecialFolder.Junk);
                        }
                        else if (model.FolderName == "All")
                        {
                            model.FolderName = "All Mail";
                            mail = client.GetFolder(SpecialFolder.All);
                        }
                        else if (model.FolderName == "Starred")
                        {
                            mail = client.GetFolder(SpecialFolder.Flagged);
                        }
                        mail.Open(FolderAccess.ReadWrite);
                    }
                    else
                    {
                        var toplevel =
client.GetFolder(client.PersonalNamespaces[0]);
                        if (model.FolderName == null || model.FolderName ==
"Inbox")
                        {
                        }
                        else if (model.FolderName == "Drafts")
                        {
                            mail = FindFolder(toplevel, model.FolderName);
                        }
                        else if (model.FolderName == "Sent")
                        {
                            mail = FindFolder(toplevel, model.FolderName);
                        }
                        else if (model.FolderName == "Starred")
                        {
                            mail = FindFolder(toplevel,
model.FolderName);
                        }
                        else if (model.FolderName == "Junk")
                        {
                            mail = FindFolder(toplevel, model.FolderName);
                        }
                        else if (model.FolderName == "Trash")
                        {
                            if (objSettings.IMAPHost == "
imap-mail.outlook.com")
                            {
                                mail = FindFolder(toplevel, "Deleted");
                            }
                            else
                            {
                                mail = FindFolder(toplevel,
model.FolderName);
                            }
                        }
                        else if (model.FolderName == "spam")
                        {
                            mail = FindFolder(toplevel, model.FolderName);
                        }
                        else if (model.FolderName == "Archive")
                        {
                            mail = FindFolder(toplevel, model.FolderName);
                        }
                        mail.Open(FolderAccess.ReadWrite);
                    }
                    model.Message = mail.GetMessage(Index);
                    model.FromAddress =
mail.GetMessage(Index).From.Mailboxes.FirstOrDefault();
                    var ReplyValues = ReplyFunction(model.Message,
model.FromAddress, ReplyAllFlag);
                    model.Attachments = ReplyValues.Attachments;
                    model.Subject = ReplyValues.Subject;
                    model.To = ReplyValues.To;
                    model.TextBody = ReplyValues.TextBody;
                    model.HtmlBody = ReplyValues.HtmlBody;
                    model.Date = ReplyValues.Date;
                    model.CC = ReplyValues.Cc;
                    model.BCC = ReplyValues.Bcc;
                    if (model.To.Count == 0)
                    {
                        model.ToAddress =
model.To.Mailboxes.FirstOrDefault().Address;
                        model.ToName =
model.To.Mailboxes.FirstOrDefault().Name;
                    }
                    else
                    {
                        foreach (var mailbox in model.To.Mailboxes)
                        {
                            model.ToAddress =
string.IsNullOrEmpty(model.ToAddress) ?

!string.IsNullOrEmpty(mailbox.Name) ? "\"" + mailbox.Name + "\"" + "<" +
mailbox.Address + ">" : mailbox.Address
                                                :
!string.IsNullOrEmpty(mailbox.Name) ? model.ToAddress + "," + "\"" +
mailbox.Name + "\"" + "<" + mailbox.Address + ">" : model.ToAddress + "," +
mailbox.Address;

                        }
                    }
                    client.Disconnect(true);
                }
            }
            catch (Exception ex)
            {
                TempData["Errormessage"] = ex.Message;
            }
            return View(MVC.Views.AdminLTE.Mailbox.Reply, model);
        }

      *  [HttpPost]*
        [ValidateInput(false)]
        public ActionResult Reply(EmailDetails model)
        {
            try
            {
                if (ModelState.IsValid)
                {
                 var message = new MimeMessage();
                message.From.Add(new MailboxAddress(objSettings.UserName));
                List<string> Receipent =
model.ToAddress.Split(',').ToList();
                InternetAddressList toMails = new InternetAddressList();
                InternetAddressList CCMails = new InternetAddressList();
                InternetAddressList BCCMails = new InternetAddressList();
                for (int i = 0; i < Receipent.Count; i++)
                {
                    MailboxAddress to = new
MailboxAddress(Receipent.ElementAt(i));
                    toMails.Add(to);                }
                message.To.AddRange(toMails);
                if (!string.IsNullOrEmpty(model.CCAddress))
                {
                    List<string> CCReceipent =
model.CCAddress.Split(',').ToList();

                    for (int i = 0; i < CCReceipent.Count; i++)
                    {
                        MailboxAddress CC = new
MailboxAddress(CCReceipent.ElementAt(i));
                        CCMails.Add(CC);
                    }
                }
                message.Cc.AddRange(CCMails);
                if (!string.IsNullOrEmpty(model.BCCAddress))
                {
                    List<string> BCCReceipent =
model.BCCAddress.Split(',').ToList();
                    for (int i = 0; i < BCCReceipent.Count; i++)
                    {
                        MailboxAddress BCC = new
MailboxAddress(BCCReceipent.ElementAt(i));
                        BCCMails.Add(BCC);
                    }
                }
                message.Bcc.AddRange(BCCMails);
                message.Subject = model.Subject;
                var builder = new BodyBuilder();
                builder.TextBody = model.TextBody;
                builder.HtmlBody = model.HtmlBody;
                    if (model.AttachedFiles != null)
                    {
                        foreach (var filePath in
model.AttachedFiles.Split(','))
                        {
                            if (System.IO.File.Exists(filePath))
                            {

builder.Attachments.Add(filePath);
                            }
                        }
                    }
                    message.Body = builder.ToMessageBody();
                using (var client = new MailKit.Net.Smtp.SmtpClient())
                {
                    client.ServerCertificateValidationCallback = (s, c, h,
e) => true;
                    client.Connect(objSettings.SMTPHost,
objSettings.SMTPPort, false);
                    client.Authenticate(objSettings.UserName,
objSettings.Password);
                    client.Send(message);
                    client.Disconnect(true);
                }
                if (model.AttachedFiles != null)
                {
                    foreach (var filePath in model.AttachedFiles.Split(','))
                    {
                        if (System.IO.File.Exists(filePath))
                        {
                            System.IO.File.Delete(filePath);
                        }
                    }
                }
                TempData["message"] = "Message sent successfully";
            }
            }
            catch (SmtpFailedRecipientException ex)
            {
                ex.GetBaseException();
                TempData["Errormessage"] =
ex.GetBaseException().Message.ToString();
            }
            catch (Exception ex)
            {
                TempData["Errormessage"] = ex.Message;
            }
            return RedirectToAction("Inbox", "Mailbox", new { flag =
"Inbox", Search = "", page = 1, min = 0, max = 0, prepageNo = 1 });

        }

VIEW CODE:
@model Test.Modules.AdminLTE.Mailbox.Classes.EmailDetails
@{
    ViewData["Title"] = "Reply";
}

@section Head {

    <script src="
http://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
    <script src="
http://hayageek.github.io/jQuery-Upload-File/4.0.11/jquery.uploadfile.min.js
"></script>
    <link rel="stylesheet"
href="~/Scripts/bootstrap-wysihtml5/bootstrap3-wysihtml5.min.css" />
    <link rel="stylesheet"
href="~/Scripts/fullcalendar/fullcalendar.min.css" />
    <link rel="stylesheet"
href="~/Scripts/fullcalendar/fullcalendar.print.css" media="print" />
    <link rel="stylesheet" href="~/Content/iCheck/flat/blue.css" />
    <link href="
http://hayageek.github.io/jQuery-Upload-File/4.0.11/uploadfile.css"
rel="stylesheet">
    <script src="
https://cdnjs.cloudflare.com/ajax/libs/moment.js/2.10.2/moment.min.js
"></script>
    <script
src="~/Scripts/bootstrap-wysihtml5/bootstrap3-wysihtml5.all.min.js"></script>
    @*<script src="https://cdn.ckeditor.com/4.8.0/standard/ckeditor.js
"></script>*@
    <script src="~/Scripts/fullcalendar/fullcalendar.min.js"></script>
    <script src="~/Scripts/jquery.icheck.min.js"></script>
    <script src="~/Scripts/adminlte/demo.js"></script>
}

@section ContentHeader {
    <h1>Mailbox</h1>
}

<script type="text/javascript">
    $(document).ready(function () {
        $("#fileuploader").uploadFile({
            url: "UploadFile",
            multiple: true,
            dragDrop: true,
            fileName: "file",
            showStatusAfterSuccess: true,
            showAbort: true,

            ////dragDropStr: "Drag Drop",

            ////abortStr: "Abort",

            ////cancelStr: "Cancel",

            ////doneStr: "Fail",

            onSuccess: function (files, data, xhr, pd) {
                //alert("Url" + data.url + "AttachedFiles" +
$('#AttachedFiles').val());
                //files: list of files
                //data: response from server
                //xhr : jquer xhr object
                $('#AttachedFiles').val($('#AttachedFiles').val() + ',' +
data.url);
            }

        });
    });
    function checkvalue() {
        var mystring = document.getElementById('ToAddress').value;
        if (!mystring.match(/\S/)) {
            toastr.error("<br/><button type='button' id='ConfirmYes'
class='btn clear'>OK</button>", 'Please enter atleast one Receipent?',
                {
                    "closeButton": true,
                    "debug": false,
                    "newestOnTop": true,
                    "progressBar": false,
                    "positionClass": "toast-top-center",
                    "preventDuplicates": true,
                    "onclick": null,
                    "showDuration": "300",
                    "hideDuration": "3000",
                    "timeOut": 0,
                    "extendedTimeOut": 0,
                    "tapToDismiss": false,
                    "allowHtml": true,
                    onShown: function (toast) {
                        $("#ConfirmYes").click(function () {
                            location.reload();
                        });
                    }
                });
            return false;
        } else {
            return true;
        }
    }
    function ShowDivCC() {
        $("#DivCC").show();
    }
    function ShowDivBCC() {
        $("#DivBCC").show();
    }
</script>

<div class="row">
    @using (Html.BeginForm("Reply", "Mailbox", FormMethod.Post, new { @id =
"ReplyId", enctype = "multipart/form-data", onsubmit = "return
checkvalue(this);" }))
    {
        @Html.AntiForgeryToken()
        @Html.HiddenFor(m => m.AttachedFiles)
        @Html.HiddenFor(m => m.ToName)
        <div class="col-md-3">
            <a href="~/AdminLTE/Mailbox/Inbox" class="btn btn-primary
btn-block margin-bottom">Back to Inbox</a>
            <div class="box box-solid">
                <div class="box-header with-border">
                    <h3 class="box-title">Folders</h3>
                    <div class="box-tools">
                        <button type="button" class="btn btn-box-tool"
data-widget="collapse">
                            <i class="fa fa-minus"></i>
                        </button>
                    </div>
                </div>
                <div class="box-body no-padding">
                    <ul class="nav nav-pills nav-stacked">
                        <li>
                            <a href="~/AdminLTE/Mailbox/Inbox">
                                <i class="fa fa-inbox"></i> Inbox
                                <span class="label label-primary
pull-right">12</span>
                            </a>
                        </li>
                        <li><a href="~/AdminLTE/Mailbox/Inbox?flag=Sent"><i
class="fa fa-envelope-o"></i>Sent</a></li>
                        <li><a
href="~/AdminLTE/Mailbox/Inbox?flag=Drafts"><i class="fa
fa-file-text-o"></i>Drafts</a></li>
                        <li>
                            <a href="~/AdminLTE/Mailbox/Inbox?flag=Junk"><i
class="fa fa-filter"></i>Junk</a>
                        <li><a
href="~/AdminLTE/Mailbox/Inbox?flag=Trash"><i class="fa
fa-trash-o"></i>Trash</a></li>
                    </ul>
                </div>
                
            </div>
        </div>
    

        <div class="col-md-9">
            <div class="box box-primary">
                <div class="box-header with-border">
                    <h3 class="box-title">Reply</h3>
                </div>
                
                <div class="box-body">
                    <div class="form-group">
                        @Html.TextBoxFor(m => m.ToAddress, new { @class =
"form-control", @placeholder = "To:" })
                        <span class="pull-right"><label id="LblCC"
onclick="ShowDivCC()">Cc</label> <label id="LblBCC"
onclick="ShowDivBCC()">Bcc</label></span>
                    </div>
                    <div class="form-group" hidden
id="DivCC">
                        @Html.TextBoxFor(m => m.CCAddress, new { @class =
"form-control", @placeholder = "Cc:" })
                    </div>
                    <div class="form-group" hidden
id="DivBCC">
                        @Html.TextBoxFor(m => m.BCCAddress, new { @class =
"form-control", @placeholder = "Bcc:" })
                    </div>
                    <div class="form-group">
                        @Html.TextBoxFor(m => m.Subject, new { @class =
"form-control", @placeholder = "Subject:" })
                    </div>
                    <div class="form-group">
                        @{
                            var str = "";
                            if (!(@Model.TextBody.IsEmptyOrNull()))
                            {
                                str = Model.TextBody.Replace("\n", "<br>");
                            }
                        }
                        <textarea id="TextBody" name="TextBody"
class="form-control" style="height: 300px">
                            On @Model.Date.ToString("f"), @Model.FromName
@Model.FromAddress wrote: <br />

                             @if (Model.HtmlBody.IsNullOrEmpty())
                              {
                                    @Html.Raw(str)
                              }
                              else
                              {
                                    @Html.Raw(Model.HtmlBody)
                              }
                    </textarea>
                        @*<script>
                             CKEDITOR.replace('TextBody');
                        </script>*@
                    </div>
                    <div class="form-group">
                        <div id="fileuploader">Upload</div>
                    </div>

                    @*<div class="form-group">
                            <div class="btn btn-default btn-file">
                                <i class="fa fa-paperclip"></i> Attachment
                                @Html.TextBoxFor(model => model.File, "",
new { @type = "file", @multiple = "multiple" })
                            </div>
                            <p class="help-block">Max. 32MB</p>
                        </div>*@
                </div>
                
                <div class="box-footer">
                    <div class="pull-right">
                        <button type="button" class="btn btn-default"><i
class="fa fa-pencil"></i> Draft</button>
                        <button type="submit" class="btn btn-primary"><i
class="fa fa-envelope-o"></i>Send</button>
                    </div>
                    <button type="reset" value="Cancel" class="btn
btn-default" onclick="location.href ='@Url.Action("Inbox", "Mailbox", new {
flag = "Inbox", Search = "" })';"><i class="fa
fa-times"></i>Discard</button>
                </div>
                
            </div>
            
        </div>

    
    }
</div>


<script>
    $(function () {
        //Add text editor
        $("#TextBody").wysihtml5();
    });
</script>

Regards,
Heena Attar

**Comment by jstedfast** [comment]
It was hard to follow what you were saying before and it's even harder to follow your code.

BTW, you can simplify this code snippet:

```csharp
                        foreach (var mailbox in model.To.Mailboxes)
                        {
                            model.ToAddress =
string.IsNullOrEmpty(model.ToAddress) ?

!string.IsNullOrEmpty(mailbox.Name) ? "\"" + mailbox.Name + "\"" + "<" +
mailbox.Address + ">" : mailbox.Address
                                                :
!string.IsNullOrEmpty(mailbox.Name) ? model.ToAddress + "," + "\"" +
mailbox.Name + "\"" + "<" + mailbox.Address + ">" : model.ToAddress + "," +
mailbox.Address;

                        }
```

to:

```csharp
model.ToAddress = model.To.ToString (false);
```

**Comment by jstedfast** [comment]
Are you trying to say that message.HtmlBody isn't finding the message's HTML content?

If so, please provide an example message.

You should also read the FAQ: https://github.com/jstedfast/MimeKit/blob/master/FAQ.md#MessageBody

**Comment by HeenaAttar** [comment]
Hello Sir,

Extremely sorry for late reply.Thanks,for minimizing my code snippet.Your
Mailkit library is really awesome and works smoothly with all my required
servers except that HTML conversion in Reply case.

Regards,
Heena Attar

**Comment by HeenaAttar** [comment]
Yes exactly message.HtmlBody is not finding the message's HTML content in
Reply case.As per your requirement i am sending an example message.

*Message: *

Content-Type: text/plain; charset=utf-8

<p>On Wednesday, January 31, 2018 3:14 PM, &quot;Dr.Batra&quot; wrote:</p>

<p>&nbsp;</p>

<blockquote>
<p><a href="
http://ga.funntale.com/ga/click/2-16111702-977-55765-109093-87122-8a24981df4-7f2e88c766"
target="_blank">Dr.Batra Hair Treatment</a></p>

<p><a href="
http://ga.funntale.com/ga/click/2-16111702-977-55765-109093-87122-8a24981df4-7f2e88c766"
target="_blank"><img alt="" src="
http://cdn1.cid247-img.move2inbox.pics/img923/7959/VFUSLj.jpg" /></a></p>

<p><a href="
http://ga.funntale.com/ga/click/2-16111702-977-55765-109093-87122-8a24981df4-7f2e88c766"
target="_blank"><img alt="" src="
http://cdn1.cid247-img.move2inbox.pics/img924/454/wZxUlm.jpg" /></a></p>

<p><a href="
http://ga.funntale.com/ga/click/2-16111702-977-55765-109093-87122-8a24981df4-7f2e88c766"
target="_blank"><img alt="" src="
http://cdn1.cid247-img.move2inbox.pics/img923/9259/3rmB5q.jpg" /></a></p>

<p><a href="
http://ga.funntale.com/ga/click/2-16111702-977-55765-109093-87122-8a24981df4-7f2e88c766"
target="_blank"><img alt="" src="
http://cdn1.cid247-img.move2inbox.pics/img922/5646/x1RPCd.jpg" /></a></p>

<p>If you find our mail in spam Please Mark as Not Spam<br />
Mark as important | <a href="mailto:remove@postvl.com?subject=I do not want
to recieve communication" target="_blank">Report Spam</a> | <a href="mailto:
remove@postvl.com?subject=I do not want to recieve communication">
Opt-out</a> | <a href="
http://ga.funntale.com/ga/click/2-16111702-977-55765-109093-87122-8a24981df4-7f2e88c766"
target="_blank">View in Browser </a></p>
<img alt="" src="
http://ga.funntale.com/ga/open/2-16111702-977-55765-109093-7f2e88c766"
style="height:2px; width:3px" /></blockquote>

<p>&nbsp;</p>



Hope this time you get what i am trying to convey😀


Regards,
Heena Attar

**Comment by jstedfast** [comment]
Your message doesn't have any HTML - it just has plain text (at least according to the Content-Type header).

If you want to treat the content as HTML, that is up to you, but MimeKit assumes that the Content-Type header is correct.

**Comment by HeenaAttar** [comment]
I will refer below code once again and try to fix the issue.Thanks, for
your quick reply to issues.


https://github.com/jstedfast/MimeKit/blob/master/FAQ.md#MessageBody
